---
title: "Interactively Exploring Soybean Breeding Populations Using R and Shiny"
author: Susan VanderPlas, Nathan Weeks, Andrew Severin, Randy Shoemaker, Jim Specht,
  Di Cook, Michelle Graham
date: "09/03/2014"
output:
  word_document:
    fig_caption: yes
csl: statistics.csl
bibliography: soybeans.bib
---


```{r setup, echo=FALSE, include=FALSE, cache=FALSE}
rm(list=ls())
options(replace.assign=TRUE,width=70)
require(knitr)
opts_chunk$set(fig.path='figure/fig-', cache.path='cache/', fig.align='center', fig.width=5, fig.height=5, fig.show='hold', par=TRUE, cache=TRUE, concordance=TRUE, autodep=TRUE)
```
# ABSTRACT

# Rationale
<!---  Literature Review -->

In the last 90 years, soybean cultivars have seen yield increases on the order of $23.4 \text{kg ha}^{−1} \text{yr}^{−1}$ [@fox2013estimating]. Part of this genetic gain comes from on-farm improvements (changes in growing practices, soil conservation, industrial fertilizer), but much of this genetic gain is due to genetic improvements that soybean breeders and farmers have generated over years of selection. The recent advancements in next-generation sequencing will allow researchers the data to identify signatures of breeder selection. However, the challenge remains of visualizing such a large, complex dataset and making the information gained usable and publicly available. The aim of this paper is to describe the development of interactive visualization tools to display the data from these experiments in an intuitive format for other researchers, allowing them to ask novel biological questions using the dataset and visualization tools. 


## Data
<!---  Describe magnitude of data, CNV analysis steps, etc. -->
The data displayed in this applet is derived from 79 lines of soybean next-generation sequencing data. Twenty seeds from each line were acquired from ---. Seeds were planted in the USDA greenhouse at Iowa State University. Once plants reached the trifoliolate stage, leaves from up to 10 plants were pooled and genomic DNA was extracted. DNA was sent to Hudson Alpha Institute for Biotechnology for next-generation sequencing. -------- Details on read-length, coverage, etc. ------. In addition, replicated field trials were conducted on a subset of lines (30 of the 79 lines, plus ancestral varieties that were not sequenced) to measure protein, oil, yield, and other characteristics under standard growth conditions, to dissociate the effect of on-farm improvements from genetic gain [@specht1984contribution],[@fox2013estimating]. 

Following sequencing, the data for each of the 79 genotypes was assembled to the *Glycine max* version 2 genome assembly [@schmutz2010genome][@phytozome10] using [GSNAP](http://research-pub.gene.com/gmap/) (version 2013-8-31) [@wu2010fast]. [samtools](http://www.htslib.org/) [@li2009sequence] was used to convert uniquely mapping reads to the BAM format. 

<!--- Read groups were added for each soybean line and duplicate reads were removed using `AddOrReplaceReadGroups` and `MarkDuplicates` functions in [picard tools](http://picard.sourceforge.net). The resulting alignment BAM files were realigned using `IndelRealigner` function in [GATK](https://www.broadinstitute.org/gatk/) [@mckenna2010genome].  The `ReduceReads` function was used to compress the alignment files by removing non-informative and redundant reads (default parameters except for downsample_coverage=1).  ---> 

cn.mops [@cnmops] was run on sequences from 79 soybean cultivars; due to computational limitations (and to provide some internal verification), cn.mops was performed separately on each genetic feature of interest defined in the genome annotation (CDS, gene, mRNA, and exons). This modification allowed us to detect copy number variation in individual exons, as well as gene duplication events. @stuparStructure found significant CNV activity on several soybean chromosomes using 4 different cultivars; this experiment vastly expands the number of cultivars and thus can detect CNVs with more power and in more cultivars. 

The files containing the results of the CNV analysis (that is, the files which are displayed using the CNV applet) are about 1.5 GB in binary form. In addition to the files containing the CNV analysis, the annotation files and other data referenced in the applet bring the total data to display up to about 2 GB. Loading that data into memory for a single user session is very reasonable; however, formatting the data so that it can be easily accessed and manipulated interactively is much more challenging. This problem is even more acute when displaying SNP data; the compressed VCF file (which is not manipulatable or plot-able) is 12 GB, and even with strict filtration, the VCF file is still at least 2 GB. Expanding this file into a database that can be easily plotted is likely to produce a file that is at least 5 times the compressed file size. 

The methodology for analysing the biological data will be described in another paper (partial citation?); in this paper we will discuss the graphics and interactive visualization of the results. 

<!---
## Biological Data and Analysis Types
A theoretical genetic analysis starts with raw sequences of genetic information, constructed from next-generation sequencing methods. These sequences are organized by chromosome, with additional sequences called "scaffolds" which are present in the genetic material but do not fit into the chromosomal sequences in a known position. From this point, genomes are aligned to a **reference genome**, so that differences may be compared between individuals or varieties within a species. Once genomes are aligned to the reference genome, biologists can conduct several different analyses for genomic features. 


A **gene** is a nucleotide sequence that encodes a specific RNA (which may be translated into a protein) sequence. It is often modified, so that portions of the gene are removed before the final RNA sequence is constructed (these portions are known as **introns**). The remaining portions of the gene (that are present in the final RNA product) are known as **exons**. The term exon refers to both the DNA sequence within a gene and to the corresponding sequence in RNA transcripts.  Exons can include both sequences that code for amino acids (red) and untranslated sequences (grey). 

![Structure of a gene, with introns and exons. 
[@Pic-GeneStructure]][GeneStructure]

### Copy number variation

One kind of genetic feature that can be analyzed using populations of genetic information is structural variation. **Copy-number variations (CNVs)**, a form of structural variation, are alterations of a genome resulting in an abnormal number of copies of one or more sections of the DNA. CNVs can occure at the gene level, or at the exon level (so that within a single gene, there are multiple copies of a sub-sequence of that gene).

![Copy number variations arise when sections of genetic material are duplicated. [@Pic-GeneDuplication]][GeneDuplication]

In the picture below, two copies of the gene (one on each chromosome) are present in the original genome (at the top of the picture), which has a copy number of **n=2**. During chromosome replication (due to meiosis or mitosis), a section of length $b+l$ is duplicated two additional times in one of the chromosomes, resulting in a final copy number (for that segment) of **n=4**. 

![CNVs arise when genes (or sections of a gene) are duplicated during chromosome replication. [@CNVcomparativeMethods]][CNVexplanation]

### Single-Nucleotide Polymorphisms

Another genomic feature that can be analyzed at the population level is single-nucleotide mutations, called **Single-Nucleotide Polymorphisms**, or **SNPs**. These point mutations are inherited but can also occur spontaneously, providing a way to analyze population structure and inheritance patterns as well as genome-wide association (where locations of SNPs are correlated with phenotypic information). 

![SNPs are single-nucleotide mutations that may alter gene function. [@Pic-SNPdiagram]][SNPdiagram]

SNPs can be associated with phenotypic traits using genome-wide association, but they can also be used to infer relatedness between two varieties: varieties which share more SNPs are assumed to be more closely related than varieties that have only a few overlapping SNPs. 

--->

## Displaying Biological Data and Analyses Interactively

### SegAnnDB - Visualizing Breakpoints for CNV modeling

http://bioinformatics.oxfordjournals.org/content/30/11/1539

Toby Hocking's pet project - interactive visualization of breakpoints. 

### Nature Methods Overview of Genomic Data Visualization
@VisualizingGenomes

* Doesn't address CNV/SNP viewing
* Lots of single-domain tools; doesn't seem to be a single extensible web-based platform. 

### QTL Charts
@qtlcharts, [Link](http://kbroman.org/qtlcharts/)

Includes several visualizations similar to those included in the CNV app. The visualizations are fairly fixed in form, though, as they are generated using d3 code from R; shiny allows any visualization that can be created in R to be created in response to user input (including qtlcharts). It would be nice to extend the heatmap to display genetic data, though this is likely to run into data size issues not present in the [qtlcharts example](http://kbroman.org/qtlcharts/example/iplotCorr.html) because the example shown seems to be a single organism, rather than a population. 


### Interactive CNV viewer (UK lab) 
@cnvinspector, [Paper](http://www.ncbi.nlm.nih.gov/pubmed/23729504), [Site](http://www.cnvinspector.org/search.mhtml)

    * Seems to be mostly human focused
    * Doesn't intuitively display the entire chromosome
    * Integrates with other databases (as we should probably integrate with Soybase)
    * Rather unintuitive. 
    
### Animint
@animint
    * Doesn't require server backend
    * Data all has to be loaded into browser memory - unwieldly for large or extremely complex datasets
    * Allows interactivity
    * Can be used within other applications (shiny bindings, rmarkdown)

## Shiny: an extensible, interactive, web-based framework for data visualization

Explain interactivity, server backend, advantage of serving up data over the web. 


# Basic Usage

## Copy Number Variation
* Describe different tabs in the applet, explaining the CNV methodology in minimal detail? Discuss the inclusion of genetic data and phenotype information (for reference purposes). 
* Discuss design considerations to deal with data size: dplyr, preprocessing data, drawing CNV=2 as the default to reduce drawn objects

The CNV Shiny applet ([Applet Link](http://gsoja.agron.iastate.edu:3838/CNV/)) contains several different tabs, which allow users to filter, manipulate, and visualize the data in different ways. The first four of these tabs, CNV Location, Copy Number, "Search CNVs by Location", and CNV List are primarily concerned with exploring the identified copy number variants; the final three tabs (Phenotype Data, Genealogy, and Methodology) provide additional information about the soybean cultivars and the experimental methodology.

The first tab is designed to show the location of an identified copy number variant for one or more cultivars. This plot is an example of a "barcode plot", where information is provided through and color intensity and location in $x$. The grey lines in the background indicate the coverage of the input sequencing data, and the green lines indicate CNV locations for a particular cultivar. The vertical alignment of different cultivars allows researchers to compare the inheritance of different CNV sites; the applet automatically sorts varieties which are related by approximate generation to facilitate these comparisons. 
![CNV Location Tab. This tab is concerned with visualizing the number of copies of a gene segment that appear in each cultivar.][cnvlocation]

In order to get more information about the number of copies of a particular genetic region, the researcher must use the next tab, "Copy Number". This tab contains information about the number of copies of a specific region, conveyed again by shades of blue. Significant copy number variants are indicated with an open circle below the vertical line. As with the last tab, this tab also contains a barcode plot, however, instead of read depth, color intensity is used to depict the number of copies of a region contained in a particular cultivar. The default copy number is 2, corresponding to one copy of the gene on each chromosome; enriched regions have a copy number higher than two, while regions that may have been deleted have a copy number of less than two. 

![Copy Number Tab. This tab displays CNV locations, to facilitate exploration of regions which might be prone to copy number variation.][cnvcopynumber]

As with the CNV Location plot, the vertical alignment of the barcode plots for different cultivars allows for an ad-hoc comparison of relatedness between cultivars. When presented with two parent varieties and their mutual child, it is possible to see which CNVs were inherited from each parent, as well as any new CNVs that were not present in either parent. 

While visual  display of the CNV analysis results is important for drawing conclusions about the overall analysis, many researchers have specific genes of interest and may want to view tabular results as well. The next two tabs are designed to facilitate that goal along with providing a user-friendly method for exporting the data for outside analysis. 

The first tab enables researchers to search for CNVs by chromosome and position on the chromosome. Users can select a range along a chromosome and the resulting interactive table will display all CNVs identified for the chromosome and region selected. The interactive table can be searched, sorted, and filtered; in particular, the results were merged with the annotated gff3 file so that each CNV region has a corresponding GlymaID. All CNVs identified for the chromosome and range of interest can be easily downloaded for analysis. 

![Search CNVs by Location. This tab allows users to search for specific positions along a chromosome and reports a dynamic, searchable table with all CNVs (and all cultivars) in the specified region. The resulting list is downloadable for analysis offline.][cnvlistrange]

Another tab allows researchers to filter by cultivar and chromosome; this separate tab is designed to allow researchers who are only interested in certain cultivars to download a subset of the data more easily. In addition, as there is no required input for this tab, a user could easily download all identified CNVs from this tab. As before, the table is dynamic and can be easily searched, sorted, and filtered by any field shown. 

![CNV List Tab. This tab allows users to search for CNVs identified in specific chromosomes and cultivars. The tablular result is again dynamic and searchable, and can be downloaded for offline analysis. ][cnvlistfilter]

### Contextual Information

The data in this experiment consisted not only of genetic sequences but also included field trials for many of the lines that were sequenced. In addition, the ancestors and descendants of most cultivars were available; this information was provided for researchers to look for context and inheritance patterns in the CNV results.

The phenotype data is presented in three sections - the main metrics for measuring soybean production (protein, oil, and yield), by year, additional soybean characteristics (maturity, seed quality, lodging, and seed size) by year, and pairwise plots of protein, oil, and yield. These sections are shown below; all graphs can be displayed simultaneously or in separate groups (for space considerations). 

![Main metrics of interest for soybean production: yield, protein, and oil, by year. Plots are interactive, and clicking on points for which parental field trial data is present displays parental data as well][PhenotypeData1]

Plots were constructed with [animint](http://tdhock.github.io/animint/) and allow user interaction - clicking on a point provides additional context about any parents for which field trial data is available. 


![Additional soybean field trial data, by year. All plots are interactive, so if multiple sections are shown, interacting with any plot will change selections for all plots.][PhenotypeData2]

![Pairwise plots of yield, protein, and oil; these plots show the tradeoff between protein and oil content and yield.][PhenotypeData3]

In addition, genealogical information is provided for all lines which are not proprietary. This allows researchers to search for related varieties, viewing inheritance of CNVs along a cultivar's family tree. 

![Genealogy of soybean cultivars. Multiple varieties can be selected; if they are related, each selected variety will be highlighted in all displayed trees.][GenealogyTab]

The genealogy plots are generated on the fly; the number of generations displayed can be altered (with more generations, it becomes more difficult to see all ancestors and descendants due to exponential increases in the number of varieties that are displayed). 

* Include example video of Michelle using the applet.

## Single Nucleotide Polymorphisms
* Describe tabs, data workflow

* Include example video of Michelle using the applet

* [Applet Link](http://gsoja.agron.iastate.edu:3838/SNP/)

## Genealogy and Phenotype Reference 
* Describe spin-off of CNV applet - genealogy of (most) soybean varieties

* [Applet Link](http://gsoja.agron.iastate.edu:3838/Genealogy/)

## Under the HOOD

### CNV Applet

* Preprocessing data step
* Reducing the number of objects in a plot
* Using javascript tables to do filtration

### SNP Applet

* Preprocessing data (VCF -> flat file, removing rows that don't correspond to SNPs that are present)
* Limiting the number of SNPs that can be viewed (for perceptual AND computational reasons)
* ?

### Genealogy Applet

* Recursive search algorithm
* phyViz package

## How does this slot into a different database like soybase?


# References

<!--- Picture References --->

[cnvlocation]:CNVLocation.png
[cnvlistfilter]:CNVListFilter.png
[cnvlistrange]:CNVListRange.png
[cnvcopynumber]:CNVCopyNumber.png
[GeneStructure]:GeneStructure.png
[GeneDuplication]:GeneDuplication.png
[CNVexplanation]:CNVexplanation.png
[PhenotypeData1]:PhenotypeData1.png
[PhenotypeData2]:PhenotypeData2.png
[PhenotypeData3]:PhenotypeData3.png
[GenealogyTab]:GenealogyTab.png
[SNPdiagram]:SNPdiagram.png

