---
title: "Interactively Exploring Soybean Breeding Populations Using R and Shiny"
author: Susan VanderPlas, Nathan Weeks, Andrew Severin, Randy Shoemaker, Jim Specht,
  Di Cook, Michelle Graham
csl: statistics.csl
output:
  word_document:
    fig_caption: yes
bibliography: soybeans.bib
---


```{r setup, echo=FALSE, include=FALSE, cache=FALSE}
rm(list=ls())
options(replace.assign=TRUE,width=70)
require(knitr)
opts_chunk$set(fig.path='figure/fig-', cache.path='cache/', fig.align='center', fig.width=5, fig.height=5, fig.show='hold', par=TRUE, cache=TRUE, concordance=TRUE, autodep=TRUE)
```
# ABSTRACT

# Rationale
<!---  Literature Review -->

In the last 90 years, soybean cultivars have seen yield increases on the order of $23.4 \text{kg ha}^{−1} \text{yr}^{−1}$ [@fox2013estimating]. Part of this genetic gain comes from on-farm improvements (changes in growing practices, soil conservation, industrial fertilizer), but much of this genetic gain is due to genetic improvements that soybean breeders and farmers have generated over years of selection. The recent advancements in next-generation sequencing will allow researchers the data to identify signatures of breeder selection. However, the challenge remains of visualizing such a large, complex dataset and making the information gained usable and publicly available. This paper describes interactive visualization tools to display the data from different sources so that soybean researchers can explore genetic signatures, genealogy and phenotypic changes. 

## Data
<!---  Describe magnitude of data, CNV analysis steps, etc. -->
Data comes from multiple sources: next-generation sequencing, large field trials, and genealogy literature. The sequencing data is derived from 79 lines of soybean next-generation sequencing data. Twenty seeds from each line were acquired from ---. Seeds were planted in the USDA greenhouse at Iowa State University. Once plants reached the trifoliolate stage, leaves from up to 10 plants were pooled and genomic DNA was extracted. DNA was sent to Hudson Alpha Institute for Biotechnology for next-generation sequencing. 

DNA sequencing libraries were prepared using TruSeq DNA sample prep and NuGENs unamplified prep kits (Illumina Inc., San Diego, CA and NuGEN Technologies Inc., San Carlos, CA).  All samples were run using Illumina V3 chemistry and 2x101bp on an Illumina 2500 instrument at Hudson Alpha Institute for Biotechnology.  The average coverage depth per line was 23x based on the total bases sequenced divided by the number of lines.


The replicated field trials were conducted on 30 of the 79 lines, and another 138 ancestral varieties that were not sequenced, to measure protein, oil, yield, and several other characteristics, under standardized growth conditions, to dissociate the effect of on-farm improvements from genetic gain [@specht1984contribution],[@fox2013estimating]. 


Following sequencing, the data for each of the 79 genotypes were aligned to the Glycine max version 2 genome assembly [@schmutz2010genome] [@phytozome10] using GSNAP (version 2013-8-31) [@wu2010fast] with default parameters except for the number of mismatches was set to 10% of the read to allow for read alignments of highly divergent soybean sequences to the Williams 82 line.  samtools[@li2009sequence] was used to convert uniquely mapping reads to the BAM format. Picard-tools (version 1.8.1) was used to add a read group to each line (AddOrReplaceReadGroups).  Optical duplicates were removed using picard-tools (MarkDuplicates). Reads were realigned using the RealignerTargetCreator and IndelRealigner functions in GATK (McKenna et al. 2010).
<!--- Read groups were added for each soybean line and duplicate reads were removed using `AddOrReplaceReadGroups` and `MarkDuplicates` functions in [picard tools](http://picard.sourceforge.net). The resulting alignment BAM files were realigned using `IndelRealigner` function in [GATK](https://www.broadinstitute.org/gatk/) [@mckenna2010genome].  The `ReduceReads` function was used to compress the alignment files by removing non-informative and redundant reads (default parameters except for downsample_coverage=1).  ---> 

cn.mops [@cnmops] was run on sequences from 79 soybean cultivars; due to computational limitations (and to provide some internal verification), cn.mops was performed separately on each genetic feature of interest defined in the genome annotation (CDS, gene, mRNA, and exons). This modification allowed us to detect copy number variation in individual exons, as well as gene duplication events. @stuparStructure found significant CNV activity on several soybean chromosomes using 4 different cultivars; this experiment vastly expands the number of cultivars and thus can detect CNVs with more power and in more cultivars. 

The files containing the results of the CNV analysis (that is, the files which are displayed using the CNV applet) are about 1.5 GB in binary form. In addition to the files containing the CNV analysis, the annotation files and other data referenced in the applet bring the total data to display up to about 2 GB. Loading that data into memory for a single user session is very reasonable; however, formatting the data so that it can be easily accessed and manipulated interactively is much more challenging. This problem is even more acute when displaying SNP data; the compressed VCF file (which is not manipulatable or plot-able) is 12 GB, and even with strict filtration, the VCF file is still at least 2 GB. Expanding this file into a database that can be easily plotted is likely to produce a file that is at least 5 times the compressed file size. 

The methodology for analysing the biological data will be described in another paper (partial citation?); in this paper we will discuss the graphics and interactive visualization of the results. 

<!---
## Biological Data and Analysis Types
A theoretical genetic analysis starts with raw sequences of genetic information, constructed from next-generation sequencing methods. These sequences are organized by chromosome, with additional sequences called "scaffolds" which are present in the genetic material but do not fit into the chromosomal sequences in a known position. From this point, genomes are aligned to a **reference genome**, so that differences may be compared between individuals or varieties within a species. Once genomes are aligned to the reference genome, biologists can conduct several different analyses for genomic features. 


A **gene** is a nucleotide sequence that encodes a specific RNA (which may be translated into a protein) sequence. It is often modified, so that portions of the gene are removed before the final RNA sequence is constructed (these portions are known as **introns**). The remaining portions of the gene (that are present in the final RNA product) are known as **exons**. The term exon refers to both the DNA sequence within a gene and to the corresponding sequence in RNA transcripts.  Exons can include both sequences that code for amino acids (red) and untranslated sequences (grey). 

![Structure of a gene, with introns and exons. 
[@Pic-GeneStructure]][GeneStructure]

### Copy number variation

One kind of genetic feature that can be analyzed using populations of genetic information is structural variation. **Copy-number variations (CNVs)**, a form of structural variation, are alterations of a genome resulting in an abnormal number of copies of one or more sections of the DNA. CNVs can occure at the gene level, or at the exon level (so that within a single gene, there are multiple copies of a sub-sequence of that gene).

![Copy number variations arise when sections of genetic material are duplicated. [@Pic-GeneDuplication]][GeneDuplication]

In the picture below, two copies of the gene (one on each chromosome) are present in the original genome (at the top of the picture), which has a copy number of **n=2**. During chromosome replication (due to meiosis or mitosis), a section of length $b+l$ is duplicated two additional times in one of the chromosomes, resulting in a final copy number (for that segment) of **n=4**. 

![CNVs arise when genes (or sections of a gene) are duplicated during chromosome replication. [@CNVcomparativeMethods]][CNVexplanation]

### Single-Nucleotide Polymorphisms

Another genomic feature that can be analyzed at the population level is single-nucleotide mutations, called **Single-Nucleotide Polymorphisms**, or **SNPs**. These point mutations are inherited but can also occur spontaneously, providing a way to analyze population structure and inheritance patterns as well as genome-wide association (where locations of SNPs are correlated with phenotypic information). 

![SNPs are single-nucleotide mutations that may alter gene function. [@Pic-SNPdiagram]][SNPdiagram]

SNPs can be associated with phenotypic traits using genome-wide association, but they can also be used to infer relatedness between two varieties: varieties which share more SNPs are assumed to be more closely related than varieties that have only a few overlapping SNPs. 

--->


# Basic Usage

There are three apps that have been constructed to address different components of the analysis: CNV, SNP, genealogy. The apps can be accessed from the [http://www.soybase.org](http://www.soybase.org). 

## Copy Number Variation
* Describe different tabs in the applet, explaining the CNV methodology in minimal detail? Discuss the inclusion of genetic data and phenotype information (for reference purposes). 
* Discuss design considerations to deal with data size: dplyr, preprocessing data, drawing CNV=2 as the default to reduce drawn objects

The CNV Shiny applet ([http://gsoja.agron.iastate.edu:3838/CNV/](http://gsoja.agron.iastate.edu:3838/CNV/)) contains several different tabs, which allow users to filter, manipulate, and visualize the data in different ways. Four of these tabs: CNV Location, Copy Number, "Search CNVs by Location", and CNV List, are primarily concerned with exploring the identified copy number variants; the final three tabs (Phenotype Data, Genealogy, and Methodology) provide additional information about the soybean cultivars and the experimental methodology. 

An initial tab provides an interactive overview of the distribution of identified CNVs by chromosome and variety. The plots shown in this tab are interactive, utilizing mouseovers to identify varieties and chromosomes, and clicks to select chromosomes or varieties to display additional information. This tab is designed to encourage the user to identify areas which may be interesting to investigate using other tabs. 

![CNV Overview. Upon initial loading, this tab displays an overview of the distribution of CNVs for each chromosome and variety. Users can interact with the plots directly through mouseovers, which provide more information, and clicks, which change plot appearance.][cnvoverview]

The second tab is designed to show the location of an identified copy number variant for one or more cultivars. This plot is an example of a "barcode plot", where information is provided through and color intensity and location in $x$. The grey lines in the background indicate the coverage of the input sequencing data, and the green lines indicate CNV locations for a particular cultivar. The vertical alignment of different cultivars allows researchers to compare the inheritance of different CNV sites; the applet automatically sorts varieties which are related by approximate generation to facilitate these comparisons. 
![CNV Location Tab. This tab is concerned with visualizing the number of copies of a gene segment that appear in each cultivar.][cnvlocation]

In order to get more information about the number of copies of a particular genetic region, the researcher must use the next tab, "Copy Number". This tab contains information about the number of copies of a specific region, conveyed again by shades of blue. Significant copy number variants are indicated with an open circle below the vertical line. As with the last tab, this tab also contains a barcode plot, however, instead of read depth, color intensity is used to depict the number of copies of a region contained in a particular cultivar. The default copy number is 2, corresponding to one copy of the gene on each chromosome; enriched regions have a copy number higher than two, while regions that may have been deleted have a copy number of less than two. 

![Copy Number Tab. This tab displays CNV locations, to facilitate exploration of regions which might be prone to copy number variation.][cnvcopynumber]

As with the CNV Location plot, the vertical alignment of the barcode plots for different cultivars allows for an ad-hoc comparison of relatedness between cultivars. When presented with two parent varieties and their mutual child, it is possible to see which CNVs were inherited from each parent, as well as any new CNVs that were not present in either parent. 

While visual  display of the CNV analysis results is important for drawing conclusions about the overall analysis, many researchers have specific genes of interest and may want to view tabular results as well. The next two tabs are designed to facilitate that goal along with providing a user-friendly method for exporting the data for outside analysis. 

The first set of tabular results enables researchers to search for CNVs by chromosome and position on the chromosome. Users can select a range along a chromosome and the resulting interactive table will display all CNVs identified for the chromosome and region selected. The interactive table can be searched, sorted, and filtered; in particular, the results were merged with the annotated gff3 file so that each CNV region has a corresponding GlymaID. All CNVs identified for the chromosome and range of interest can be easily downloaded for analysis. 

![Search CNVs by Location. This tab allows users to search for specific positions along a chromosome and reports a dynamic, searchable table with all CNVs (and all cultivars) in the specified region. The resulting list is downloadable for analysis offline.][cnvlistrange]

Another tab allows researchers to filter by cultivar and chromosome; this separate tab is designed to allow researchers who are only interested in certain cultivars to download a subset of the data more easily. In addition, as there is no required input for this tab, a user could easily download all identified CNVs directly by not providing any filtration. As before, the table is dynamic and can be easily searched, sorted, and filtered by any field shown. GlymaIDs for identified CNVs link to Soybase, integrating the applet into the pre-existing database infrastructure.

![CNV List Tab. This tab allows users to search for CNVs identified in specific chromosomes and cultivars. The tablular result is again dynamic and searchable, and can be downloaded for offline analysis. ][cnvlistfilter]

### Contextual Information

The data in this experiment consisted not only of genetic sequences but also included field trials for many of the lines that were sequenced. In addition, the ancestors and descendants of most cultivars were available; this information was provided for researchers to look for context and inheritance patterns in the CNV results.

The phenotype data is presented in three sections - the main metrics for measuring soybean production (protein, oil, and yield), by year, additional soybean characteristics (maturity, seed quality, lodging, and seed size) by year, and pairwise plots of protein, oil, and yield. These sections are shown below; all graphs can be displayed simultaneously or in separate groups (for space considerations). 

![Main metrics of interest for soybean production: yield, protein, and oil, by year. Plots are interactive, and clicking on points for which parental field trial data is present displays parental data as well][PhenotypeData1]

Plots were constructed with [animint](http://tdhock.github.io/animint/) and allow user interaction - clicking on a point provides additional context about any parents for which field trial data is available. 


![Additional soybean field trial data, by year. All plots are interactive, so if multiple sections are shown, interacting with any plot will change selections for all plots.][PhenotypeData2]

![Pairwise plots of yield, protein, and oil; these plots show the tradeoff between protein and oil content and yield.][PhenotypeData3]

In addition, genealogical information is provided for all lines which are not proprietary. This allows researchers to search for related varieties, viewing inheritance of CNVs along a cultivar's family tree. 

![Genealogy of soybean cultivars. Multiple varieties can be selected; if they are related, each selected variety will be highlighted in all displayed trees.][GenealogyTab]

The genealogy plots are generated on the fly; the number of generations displayed can be altered (with more generations, it becomes more difficult to see all ancestors and descendants due to exponential increases in the number of varieties that are displayed). 

* Include example video of Michelle using the applet.

## Single Nucleotide Polymorphisms
* Include example video of Michelle using the applet

* [http://gsoja.agron.iastate.edu:3838/SNP/](http://gsoja.agron.iastate.edu:3838/SNP/)

The SNP Shiny applet aims to provide an interface for biologists to search for SNPs, view SNPs along the genome, and examine the distribution of SNPs along the genome. A preliminary analysis, which heavily filtered SNPs using several criteria, still identified more than one million SNP sites, with nearly 20 million total SNPs in 79 different varieties. This data is much too large to visualize in detail with static plots, as the data must be heavily summarized for the plots to be readable. The applet aims to provide a "zoom and filter" approach to the data; displaying summaries while providing more detail and searchable tabular results. Like the CNV applet, the SNP applet has several tabs which provide data at different levels of detail. 

![Aggregate SNP Browser. This tab displays the proportion of varieties with SNPs at each position along the chromosome.][SNPAggregateBrowser]

The first tab provides a nucleotide level view of the frequency of each SNP. Up to 50 SNPs can be viewed in the same window, and researchers can identify sites with rare SNPs or sites with large amounts of variability. The chromosome and position along the chromosome can be changed through several different methods: researchers can enter a gene ID, and the plot will update to show SNPs at the start of a gene; alternately, researchers can choose a chromosome and a position manually. Researchers can also move upstream and downstream along the genome directly, using buttons located beneath the plot. A table below the plot displays gene IDs matching the string entered; this allows researchers to search for partial string matches, or list all genes on a specific chromosome to explore. 


![SNP Counts by Glyma ID. This tab allows researchers to explore which Glyma IDs have SNPs (and how many varieties have a SNP along that gene), and then further explore the number of varieties with each SNP for a chosen gene.][SNPCountsByID]

The next tab allows researchers to see which IDs have a large number of SNPs (or SNPs that are extremely common). Initially, results can be filtered by chromosome, and a table displaying the number of SNPs associated with each gene ID and the number of varieties with SNPs along that gene is provided. An additional breakdown of SNP locations within a single gene shows the number of varieties with a SNP at a specific position. These tables are interactive and can be sorted, searched, and filtered by any field shown. 

![SNP Counts by Glyma ID. Searching by a specific Glyma ID shows all of the positions along the chromosome (but within the gene) that have identified SNPs, along with the number of varieties having a SNP at each position.][SNPCountsByID1]

In addition to facilitating browsing by position along the genome, SNPs for a specific gene can also be visualized by variety. If no varieties are specified, SNPs will be displayed for the first 10 alphabetical varieties with SNPs identified along the specified gene. This type of plot allows researchers to look at the distribution of SNPs within a population; varieties with a certain genotype can be visually grouped, and could be potentially associated with known phenotypes. As with the CNV applet, SNPs with a corresponding GlymaID link to the relevant Soybase entry for that ID, integrating the applet with existing Soybase infrastructure. 

![Variety-Level SNP Browser. Searching by a specific Glyma ID shows all of the identified SNPs along a gene, and the identified frequency of the alternate allele (compared to the reference) for up to 10 varieties.][SNPVarietyBrowser]

For instance, Glyma.18G000100.Wm82.a2.v1 has several SNPs that are common in older varieties descended from CNS, A.K.,  and Mukden; S-100 and Hawkeye are both heterozygous for several SNPs in that gene as well. Underneath the plot, an interactive table provides estimated allele frequency, gene state, and approximate allele count for each displayed variety. 

![Variety-Level SNP Browser. A table underneath the plot provides details, including estimated allele frequency and estimated gene state.][SNPVarietyBrowser1]

For a more high-level view, another tab provides a density plot, by chromosome and variety, showing the distribution of SNPs along a chromosome. This allows researchers to visually identify regions with numerous SNPs for comparison to QTL data and identification of regions of the genome which might merit further exploration. 

![SNP Density. This plot shows the density of identififed alternate alleles for a number of varieties.][SNPDensity]

In the plot above, Harosoy and Corsoy are obviously related (Corsoy is Harosoy's direct descendant), as they have extremely similar alternate allele densities at the end of the chromosome, however, Harosoy has several regions of SNP activity in the middle of chromosome 20 that are not present in Corsoy. Zane, a descendant of all of the other displayed varieties, has two extremely dense regions of SNPs that have been concentrated through inbreeding, and relatively few SNPs in other locations on chromosome 20. 

![SNP Kinship. This plot shows the kinship between different lines based on SNP dissimilarity and generational distance. Users can click on a specific square of the heatmap and the other plots will change accordingly.][SNPKinship]

* Add information on SNP Kinship calculation and Kevin Bacon distance

![SNP Kinship. Underneath the two heatmaps, there are three additional plots (also linked with the heatmaps) which provide additional information showing the shortest path between two varieties, the relationship between SNP dissimilarity and generational distance, and the difference in yield between the two varieties (if both varieties were utilized in the field trial associated with the sequencing study).][SNPKinship1]

The first plot shown above provides the genealogical link between any two related varieties; the shortest path was calculated using the phyViz package. The second plot shows the relationship between generational distance and SNP distance; as the number of generations increases, the SNP distance increases (the maximum SNP distance is 2). The third plot shows the SNP distance between the two varieties as a function of SNP Distance; lower SNP distance is associated with a lower yield difference, but higher SNP distance does not necessarily indicate a large difference in yield. Clicking on any point in the two scatterplots will highlight the associated varieties on the linked heatmaps, and the shortest path between the two varieties will also be displayed. 

# Materials and Methods
<!--- 
## Other Visualization Software 
 * SegAnnDB - Visualizing Breakpoints for CNV modeling - http://bioinformatics.oxfordjournals.org/content/30/11/1539 - Toby Hocking's pet project - interactive visualization of breakpoints. 
 * Nature Methods Overview of Genomic Data Visualization    
@VisualizingGenomes
    * Doesn't address CNV/SNP viewing
    * Lots of single-domain tools; doesn't seem to be a single extensible web-based platform. 
 * QTL Charts     
@qtlcharts, [Link](http://kbroman.org/qtlcharts/)    
Includes several visualizations similar to those included in the CNV app. The visualizations are fairly fixed in form, though, as they are generated using d3 code from R; shiny allows any visualization that can be created in R to be created in response to user input (including qtlcharts). It would be nice to extend the heatmap to display genetic data, though this is likely to run into data size issues not present in the [qtlcharts example](http://kbroman.org/qtlcharts/example/iplotCorr.html) because the example shown seems to be a single organism, rather than a population. 
 * Interactive CNV viewer (UK lab)     
@cnvinspector, [Paper](http://www.ncbi.nlm.nih.gov/pubmed/23729504), [Site](http://www.cnvinspector.org/search.mhtml)    
    * Seems to be mostly human focused
    * Doesn't intuitively display the entire chromosome
    * Integrates with other databases (as we should probably integrate with Soybase)
    * Rather unintuitive. 

 
## Animint    
@animint    
Not specific to biology, but designed with bioinformatics visualization as a primary application.

  * Doesn't require server backend
  * Data all has to be loaded into browser memory - unwieldly for large or extremely complex datasets
  * Allows interactivity
  * Can be used within other applications (shiny bindings, rmarkdown)
 --->
## Shiny: an extensible, interactive, web-based framework for data visualization
Displaying large data sets over the web using Shiny applets and interactive javascript plots requires that we carefully consider the computational and bandwidth requirements of each section and feature of the applet. This section describes the problems we encountered during the applet construction process, and the solutions we found to be most helpful in resolving the issue. 

One issue common to both applets was the difference between data shared between sessions and data loaded within each session. When multiple users are connected to the same applet, if data is shared between the users, the second user to connect has a nearly-instantaneous applet loading time (compared to a several-second period as the data is loaded if each user must re-load the data). Sharing data between sessions does require that no session modify the data (so all operations on the data set must be stored in temporary variables and computed on the fly). This adds some additional computational burden, but reduces loading time for additional users who connect to the applet. 

## CNV Applet

* Add stats on size of data, etc.

All of the data displayed using the CNV applet has been carefully constructed to provide the relevant information (copy number, CNV location, glymaID, etc.) while minimizing the amount of data that must be transferred. For instance, the CNV location plot uses binned read count data, and the copy number plot uses a default background of CN=2; both of these modifications are designed to reduce the number of rectangles that have to be drawn (and thus, reduce the amount of time it takes to update the plot). The binned read count data and the copy number variant tables are pre-computed in a processing script that is run after the cn.mops program identifies CNVs. 

The genealogy search function is written as a recursive function that is extremely fast; the generational limits speed the database search up even more (and also reduce the number of things plotted, for visibility and speed). 

DataTables, a javascript library, is used to display reported CNV information. This library provides search, sort, and filtration features, reducing the need to communicate with the server for these operations. Certain operations (filtering by variety and chromosome) do utilize the server, but this step is necessary for creating the barcode plots, and so it is only natural to require table output that is consistent with the plots. The linked plots in the Phenotype data tab are also rendered using the d3 javascript library (they were compiled into d3 code using the animint R package [@animint]); the interactive plots are designed to minimize load time by reading in only the necessary data during the initial rendering. 

These modifications helped to make the CNV applet much more responsive and user-friendly. Modifying the CNV applet to streamline the loading process was extremely good training for the SNP applet, which has magnitudes more data and required even more pre-processing.

## SNP Applet
The SNP applet contains several tabs which display different aspects of a single underlying data file (the Kinship tab incorporates some additional data as well). The underlying data file contains the (heavily filtered) SNPs for each of 79 varieties; when compressed, it is 1.18 GB, but when expanded to a tabular format which can be plotted easily, the file grows to 5.2 GB. On a sufficiently powerful computer, R can handle data of this magnitude, but each operation is proportionately slower. To optimize applet performance, we first removed all lines from the expanded table which did not correspond to SNPs for a combination of position and Variety, as well as SNPs on scaffolds. This reduced the dataset from 87 million rows to 20 million rows. The resulting dataset can now be loaded on most modern computers, and critical operations (sort, summarize, filter) are now much more feasible. 

In order to minimize the number of operations that have to be completed while the applet is running, several summary data files are pre-computed. The aggregate browser and the SNP Density tab both display summaries of data (over variety or position); these summaries are pre-computed and then filtered in response to user input. The SNP Counts tab similarly relies on filtering aggregated data; its purpose is to bridge the gap between the aggregate SNP browser and the glymaIDs necessary to use the variety-level SNP browser. 

In addition to pre-computed files, the applet is carefully designed so that user queries do not return enough results to overwhelm the server (or the client). The aggregate SNP browser, for instance, is limited to 50 SNPs at a time; the variety-level SNP browser can only display 10 cultivars and a single GlymaID. These modifications are essential both for the responsiveness of the server and to prevent visual overload. The SNP counts tab utilizes a combination of server-side filtration using user inputs, and client-side filtration and sorting using DataTables.js. This combination provides responsiveness through client-side interaction without overwhelming the browser's rendering system. 

Even with these modifications, we found that the applet is computationally intensive enough to justify purchasing the pro version of Shiny Server, which has threaded performance and other features that allow the applet to run more smoothly for multiple simultaneous users. 


# Discussion
## How does this slot into a different database like soybase?
The two applets have been constructed to integrate with existing databases; at this time, the applet links users to Soybase, but functionality exists to link to the applets (with specific input specified in the URL), providing the ability to direct users to the applet from other sites as well as providing additional information located in other databases within the Shiny applet. 

> Might be useful to link to GRIN in the genealogy tabs...?

# References
McKenna, A. et al., 2010. The Genome Analysis Toolkit: A MapReduce framework for analyzing next-generation DNA sequencing data. Genome research, 20(9), pp.1297–1303.
<!--- Picture References --->

[cnvoverview]:CNVOverview.png
[cnvlocation]:CNVLocation.png
[cnvlistfilter]:CNVListFilter.png
[cnvlistrange]:CNVListRange.png
[cnvcopynumber]:CNVCopyNumber.png
[GeneStructure]:GeneStructure.png
[GeneDuplication]:GeneDuplication.png
[CNVexplanation]:CNVexplanation.png
[PhenotypeData1]:PhenotypeData1.png
[PhenotypeData2]:PhenotypeData2.png
[PhenotypeData3]:PhenotypeData3.png
[GenealogyTab]:GenealogyTab.png
[SNPdiagram]:SNPdiagram.png
[SNPAggregateBrowser]:SNPAggregateBrowser.png
[SNPCountsByID]:SNPCountsByID1.png
[SNPCountsByID1]:SNPCountsByID.png
[SNPVarietyBrowser]:SNPVarietyBrowser.png
[SNPVarietyBrowser1]:SNPVarietyBrowser1.png
[SNPDensity]:SNPDensity.png
[SNPKinship]:SNPKinship.png
[SNPKinship1]:SNPKinship1.png
