---
title: "Interactive Display of Population Soybean Genomic Data with Shiny"
author: "Susan VanderPlas, Nathan Weeks, Jim Specht, Di Cook, Michelle Graham"
date: "09/03/2014"
output:
  word_document:
    fig_caption: yes
---


```{r setup, echo=FALSE, include=FALSE, cache=FALSE}
rm(list=ls())
options(replace.assign=TRUE,width=70)
require(knitr)
opts_chunk$set(fig.path='figure/fig-', cache.path='cache/', fig.align='center', fig.width=5, fig.height=5, fig.show='hold', par=TRUE, cache=TRUE, concordance=TRUE, autodep=TRUE)
```
# ABSTRACT

# Interactive Display of Biological Data
<!---  Literature Review -->

### Nature Methods Overview of Genomic Data Visualization
[Link](http://mipt.ru//dbmp/upload/d0d/Nielsen-arphlf450u6.pdf)

* Doesn't address CNV/SNP viewing
* Lots of single-domain tools; doesn't seem to be a single extensible web-based platform. 

### QTL Charts
[Link](http://kbroman.org/qtlcharts/)

Includes several visualizations similar to those included in the CNV app. The visualizations are fairly fixed in form, though, as they are generated using d3 code from R; shiny allows any visualization that can be created in R to be created in response to user input (including qtlcharts). It would be nice to extend the heatmap to display genetic data, though this is likely to run into data size issues not present in the [qtlcharts example](http://kbroman.org/qtlcharts/example/iplotCorr.html) because the example shown seems to be a single organism, rather than a population. 


### Interactive CNV viewer (UK lab) 
[Paper](http://www.ncbi.nlm.nih.gov/pubmed/23729504) [Site](http://www.cnvinspector.org/search.mhtml)

    * Seems to be mostly human focused
    * Doesn't intuitively display the entire chromosome
    * Integrates with other databases (as we should probably integrate with Soybase)
    * Rather unintuitive. 

## Shiny: an extensible, interactive, web-based framework for data visualization

Explain interactivity, server backend, advantage of serving up data over the web. 

# Data
<!---  Describe magnitude of data, CNV analysis steps, etc. -->
The data displayed in this applet is derived from 79 lines of soybean next-generation sequencing data. The files containing the results of the CNV analysis (that is, the files which are displayed using the CNV applet) are about 1.5 GB in binary form. In addition to the files containing the CNV analysis, the annotation files and other data referenced in the applet bring the total data to display up to about 2 GB. Loading that data into memory for a single user session is very reasonable; however, formatting the data so that it can be easily accessed and manipulated interactively is much more challenging. This problem is even more acute when displaying SNP data; the compressed VCF file (which is not manipulatable or plot-able) is 12 GB, and even with strict filtration, the VCF file is still at least 2 GB. Expanding this file into a database that can be easily plotted is likely to produce a file that is at least 5 times the compressed file size. 

The methodology for analysing the biological data will be described in another paper (partial citation?); in this paper we will discuss the graphics and interactive visualization of the results. 



# Basic Usage

## Copy Number Variation
* Describe different tabs in the applet, explaining the CNV methodology in minimal detail? Discuss the inclusion of genetic data and phenotype information (for reference purposes). 
* Discuss design considerations to deal with data size: dplyr, preprocessing data, drawing CNV=2 as the default to reduce drawn objects

The CNV Shiny applet ([Applet Link](http://gsoja.agron.iastate.edu:3838/CNV/)) contains several different tabs, which allow users to filter, manipulate, and visualize the data in different ways. The first four of these tabs, CNV Location, Copy Number, "Search CNVs by Location", and CNV List are primarily concerned with exploring the identified copy number variants; the final three tabs (Phenotype Data, Genealogy, and Methodology) provide additional information about the soybean cultivars and the experimental methodology.

![CNV Location Tab. This tab is concerned with visualizing the number of copies of a gene segment that appear in each variety.][cnvlocation]


![Copy Number Tab. This tab displays CNV locations, to facilitate exploration of regions which might be prone to copy number variation.][cnvcopynumber]

![Search CNVs by Location. This tab allows users to search for specific positions along a chromosome and reports a dynamic, searchable table with all CNVs (and all cultivars) in the specified region. The resulting list is downloadable for analysis offline.][cnvlistrange]

![CNV List Tab. This tab allows users to search for CNVs identified in specific chromosomes and cultivars. The tablular result is again dynamic and searchable, and can be downloaded for offline analysis. ][cnvlistfilter]

* Include example video of Michelle using the applet.

## Single Nucleotide Polymorphisms
* Describe tabs, data workflow

* Include example video of Michelle using the applet

* [Applet Link](http://gsoja.agron.iastate.edu:3838/SNP/)

## Genealogy and Phenotype Reference 
* Describe spin-off of CNV applet - genealogy of (most) soybean varieties

* [Applet Link](http://gsoja.agron.iastate.edu:3838/Genealogy/)

[cnvlocation]:CNVLocation.png
[cnvlistfilter]:CNVListFilter.png
[cnvlistrange]:CNVListRange.png
[cnvcopynumber]:CNVCopyNumber.png

## Under the HOOD

### CNV Applet

* Preprocessing data step
* Reducing the number of objects in a plot
* Using javascript tables to do filtration

### SNP Applet

* Preprocessing data (VCF -> flat file, removing rows that don't correspond to SNPs that are present)
* Limiting the number of SNPs that can be viewed (for perceptual AND computational reasons)
* ?

### Genealogy Applet

* Recursive search algorithm
* phyViz package

## How does this slot into a different database like soybase?
